packages/toolpad-app/src/server/localMode.ts

```diff
@@ -989,11 +989,16 @@ class ToolpadProject {
 
   private pendingVersionCheck: Promise<VersionInfo> | undefined;
 
-  constructor(root: string, options: Partial<ToolpadProjectOptions>) {
+  constructor(
+    root: string,
+    options: Partial<ToolpadProjectOptions>,
+    externalUrl: string = 'http://localhost:3000',
```

Comment:
Let's put `externalUrl` in the options, as per https://github.com/mui/mui-toolpad/issues/2648#issuecomment-1721312565

Reply:
![image](https://github.com/mui/mui-toolpad/assets/27261522/4d01af7e-b53a-44dd-ae55-129995852730)
![image](https://github.com/mui/mui-toolpad/assets/27261522/cab9df74-2177-421e-883d-297a13817843)

Follow your step, there is a problem, if it doesn't set a default value.

Reply:
I don't believe your screenshot corresponds to https://github.com/mui/mui-toolpad/issues/2648#issuecomment-1721312565

Reply:
![image](https://github.com/mui/mui-toolpad/assets/27261522/e1cd027f-1e32-4483-8e31-e2c7c8f1a54f)

you pass to `ToolpadProject`, so that we need to receive a `externalUrl`

and we can't directly pass to the `options` because  the `externalUrl` maybe is  `undefined `

![image](https://github.com/mui/mui-toolpad/assets/27261522/b5406b04-6192-427a-b3ba-8f310c1ad9bd)

Reply:
![image](https://github.com/mui/mui-toolpad/assets/27261522/c69d286d-cdbd-454b-bc4b-997c4fc828cb)

Reply:
It's being passed to the `ToolpadProject` as a `externalUrl` property of the second parameter `options`. Not as a third parameter `externalUrl`

Reply:
my bad, i'm sory. I'll fix it. lol